contracts/MechMarketplace.sol

```diff
@@ -489,7 +506,154 @@ contract MechMarketplace is IErrorsMarketplace {
+        emit RequesterHashApproved(msg.sender, hash);
+    }
+    
+    /// @dev Delivers signed requests.
+    /// @notice This function must be called by mech delivering requests.
+    /// @param requester Requester address.
+    /// @param requestDatas Corresponding set of self-descriptive opaque request data-blobs.
+    /// @param signatures Corresponding set of signatures.
+    /// @param deliveryRates Corresponding set of actual charged delivery rates for each request.
+    /// @param deliveryDatas Corresponding set of self-descriptive opaque delivery data-blobs.
+    /// @param paymentData Additional payment-related request data, if applicable.
+    function deliverMarketplaceBatchWithSignatures(
+        address requester,
+        bytes[] memory requestDatas,
+        bytes[] memory signatures,
+        bytes[] memory deliveryDatas,
+        uint256[] memory deliveryRates,
+        bytes memory paymentData
+    ) external payable {
+        // Check mech
+        address mechServiceMultisig = checkMech(msg.sender);
+
+        // TODO array length checks
+
+        // TODO check requester?
+
+        // Get number of requests
+        // Allocate set for request Ids
+        uint256[] memory requestIds = new uint256[](requestDatas.length);
+
+        uint256 totalDeliveryRate;
+        uint256 nonce = mapNonces[requester];
+
+        // Traverse all request Ids
+        for (uint256 i = 0; i < requestDatas.length; ++i) {
+            requestIds[i] = getRequestId(requester, requestDatas[i], nonce);
+
+            // Verify the signed hash against the operator address
+            _verifySignedHash(requester, bytes32(requestIds[i]), signatures[i]);
+
+            // Assign mech delivery info struct values
+            mapRequestIdDeliveries[requestIds[i]] = MechDelivery(msg.sender, msg.sender, requester, 0, deliveryRates[i]);
+
+            // Increase total rate
+            totalDeliveryRate += deliveryRates[i];
+
+            // Increase nonce
+            nonce++;
+        }
+
+        // Adjust requester nonce values
+        mapNonces[requester] = nonce;
+
+        // Increase the amount of requester delivered requests
+        mapDeliveryCounts[requester]++;
+        // Increase the amount of mech delivery counts
+        mapMechDeliveryCounts[msg.sender]++;
+        // Increase the amount of mech service multisig delivered requests
+        mapMechServiceDeliveryCounts[mechServiceMultisig]++;
+
+        // Increase mech karma that delivers the request
+        IKarma(karma).changeMechKarma(msg.sender, int256(requestDatas.length));
+
+        // Get balance tracker address
+        bytes32 mechPaymentType = IMech(msg.sender).paymentType();
+        address balanceTracker = mapPaymentTypeBalanceTrackers[mechPaymentType];
+
+        // Process payment
+        IBalanceTracker(balanceTracker).adjustRequesterMechBalances{value: msg.value}(msg.sender, requester,
+            totalDeliveryRate, paymentData);
```

Comment:
I believe we should still give a chance to pre-pay via a mech if there's not enough credit?

Reply:
I don't understand what the usecase would be. Why would the mech pre-pay? They already had the off-chain cost of doing this, why would they now pay themselves?

Reply:
If requester didn't have enough balance on Mech Marketplace, is there a possibility to relay value via the mech, such that the requester balance increases and becomes sufficient to pay for the mech work? Or we consider all balances final before executing requests and the requester either has enough balances to pay for or doesn't.

Reply:
ok just talked to Ardian, this is super silly :) I will modify it as obviously mech acts without a requester to ask for additional funds.